src/lsm/schema.zig

```diff
@@ -509,3 +520,122 @@ pub const TableData = struct {
         return schema.block_values_bytes_const(data_block)[0..used_bytes];
     }
 };
+
+/// A ManifestLog block's body is a SoA of Labels and TableInfos.
+/// Each Label/TableInfo pair is an "entry".
+// TODO Store timestamp (snapshot) in header.
+pub const ManifestLog = struct {
+    const entry_size = @sizeOf(Label) + @sizeOf(TableInfo);
+    const entry_count_max_unaligned = @divFloor(block_body_size, entry_size);
+
+    pub const entry_count_max =
+        @divFloor(entry_count_max_unaligned, @alignOf(TableInfo)) * @alignOf(TableInfo);
+
+    const labels_size_max = entry_count_max * @sizeOf(Label);
+    const tables_size_max = entry_count_max * @sizeOf(TableInfo);
+
+    comptime {
+        assert(entry_count_max > 0);
+        assert(labels_size_max % @alignOf(TableInfo) == 0);
+        assert(tables_size_max % @alignOf(TableInfo) == 0);
+
+        // Bit 7 is reserved to indicate whether the event is an insert or remove.
+        assert(constants.lsm_levels <= std.math.maxInt(u7) + 1);
+
+        assert(@sizeOf(Label) == @sizeOf(u8));
+
+        // All TableInfo's should already be 16-byte aligned because of the leading checksum.
+        const alignment = 16;
+        assert(@alignOf(TableInfo) == alignment);
+
+        // For keys { 8, 16, 24, 32 } all TableInfo's should be a multiple of the alignment.
+        // However, we still store Label ahead of TableInfo to save space on the network.
+        // This means we store fewer entries per manifest block, to gain less padding,
+        // since we must store entry_count_max of whichever array is first in the layout.
+        // For a better understanding of this decision, see schema.ManifestLog.size().
+        assert(@sizeOf(TableInfo) % alignment == 0);
+    }
+
+    pub const Label = packed struct(u8) {
+        level: u7,
+        event: enum(u1) { insert, remove },
+
+        comptime {
+            assert(@bitSizeOf(Label) == @sizeOf(Label) * 8);
+        }
+    };
+
+    /// See manifest.zig's TreeTableInfoType declaration for field documentation.
+    pub const TableInfo = extern struct {
```

Comment:
:thinking: I'm not sure what you mean?

Reply:
Sorry, GitHub ate half of the comment. When processing manifest log entries, we skip over entries for addresses we have seen “in the future”, because we know they are outdated. If addresses were a separate array, then we could skip loading other files of table info in cache for skipped over entries

Reply:
Ah, I see